src/rust/engine/ui/src/display.rs

```diff
@@ -266,13 +275,39 @@ impl EngineDisplay {
   }
 
   // Paints one screen of rendering.
-  pub fn render(&mut self) {
+  pub fn render(&mut self) -> std::result::Result<KeyboardCommand, String> {
     self.set_size();
     self.clear();
     let max_log_rows = self.get_max_log_rows();
     let rendered_count = self.render_logs(max_log_rows);
     self.render_actions(rendered_count);
-    self.flush().expect("could not flush terminal!");
+    if let Err(err) = self.flush() {
+      return Err(format!("Could not flush terminal: {}", err));
+    }
+    self.handle_stdin()
+  }
+
+  fn handle_stdin(&mut self) -> std::result::Result<KeyboardCommand, String> {
+    use termion::event::{parse_event, Event, Key};
+    //This buffer must have non-zero size because termion's `read` implementation for
+    //AsyncReader will return early without doing anything if it is of length 0.
+    //(See https://docs.rs/termion/1.5.4/src/termion/async.rs.html#69 )
+    let mut buf: [u8; 32] = [0; 32];
+
+    match self.async_stdin.read(&mut buf) {
+      Ok(0) => Ok(KeyboardCommand::None),
+      Ok(_) => {
+        let initial_byte: u8 = buf[0];
+        let mut iter = buf[1..].iter().map(|byte| Ok(*byte));
+        let event_or_err = parse_event(initial_byte, &mut iter);
```

Comment:
It doesn't look like this method will do the right thing... it will only actually see an event at the beginning of a sequence of bytes, so I think that if someone typed `aCtrl+C` fast enough, this would see the `a` and ignore the trailing bytes?

It feels like rather than reading input as bytes, this should maybe be reading using an Event iterator like https://docs.rs/termion/1.5.5/termion/input/struct.Events.html ?

Reply:
I tried using that, but the [method that creates it](https://docs.rs/termion/1.5.5/termion/input/trait.TermRead.html#tymethod.events) takes`self`, so it tries to move `self.async_stdin` into the `render` method every time it's called, which the borrow-checker prevents. If I try to create `async_stdin()` as a local variable of the `render` method it compiles but never actually detects any keystrokes.

Reply:
Hm. Not sure what to tell you there... but I'm 90% sure that this will miss events? It might be worth filing a ticket to ask them how to use that API?

Reply:
As far as I can tell from the docs, `async_stdin` sets up a mpsc channel (which I was thinking of doing manually in order to solve the problem of reading from stdin without blocking before I looked at how `async_stdin` was implemented), so it makes sense that, if we created it anew in the render method, we'd be constantly setting up and taking down the channel without giving it a chance to actually read anything. Whereas if we create it in the body of the EngineDisplay struct, it will live for the length of the EngineDisplay run and have enough time to collect input. It is annoying that we can't get an Events iterator out of this, but I think this solution is good enough for what we need to do with it. It's very possible that I'm missing something in my mental model of how unix console IO works in a raw terminal context that would lead me to a different solution.